### PR TITLE
BI-1623 [Buildbot] Create "build" function for building now a Jenkin job

### PR DESCRIPTION
**Pull request by nguyentunglam912**
Create build_jenkin command for building now a job on Jenkin
image

### CODE REVIEW

File: workspace/lib/jenkinhelper/jenkin.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      username: config.username,
+      password: config.password)
+    end
+
+    def get_item_in_queue(job_name)
+      queue_result = client.api_get_request("/job/#{path_encode job_name}")['inQueue']
+      if queue_result
+        return client.api_get_request("/job/#{path_encode job_name}")['nextBuildNumber']
+      else
+        return queue_result
```
Comment (nguyentranlogigear) [inline workspace/lib/jenkinhelper/jenkin.rb]: I think that we can update
      if queue_result
         queue_result = client.api_get_request("/job/#{path_encode job_name}")['nextBuildNumber']
      return queue_result

File: workspace/lib/lita/handlers/jenkin.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        response.reply(t('build_jenkin.building', job_name: job_name))
+        begin
+          if job_exists?(job_name)
+            if (params.nil? or params.empty?)
+              client.api_post_request("/job/#{path_encode job_name}/build", {}, true)
+            else
+              client.api_post_request("/job/#{path_encode job_name}/buildWithParameters", params, true)
+            end
+            queue_result = get_item_in_queue(job_name)
+            unless queue_result == true
```
Comment (nguyentranlogigear) [inline workspace/lib/lita/handlers/jenkin.rb]: I think that <code>== true</code> is not necessary

Comment (nguyentunglam912) [reply workspace/lib/lita/handlers/jenkin.rb]: it should be == false instead of true. I mistook this line.

**Comment by scopp** [comment]
@nguyentranlogigear can you please refactor all the references of `jenkin` to `jenkins`

**Review by scopp — CHANGES_REQUESTED** [review]
@nguyentranlogigear can you please refactor all the references of `jenkin` to `jenkins`

**Comment by nguyentunglam912** [comment]
@scopp I have refactored all the references of jenkin to jenkins.

File: workspace/lita_config.rb [hunk 3]
```diff
@@ -41,5 +41,8 @@
   config.handlers.teamcity.password         = ENV['TEAMCITY_PASSWORD'] || ''
   config.handlers.teamcity.git_uri          = ENV['GIT_URI'] || ''
   config.handlers.teamcity.python_script    = ENV['PYTHON_SCRIPT'] || ''
+  config.handlers.jenkins.username           = ENV['JENKIN_USERNAME']
+  config.handlers.jenkins.password           = ENV['JENKIN_PASSWORD']
+  config.handlers.jenkins.ip                 = ENV['JENKIN_IP']
```
Comment (nguyentranlogigear) [inline workspace/lita_config.rb]: Please help me align variable in lita_config
Update Jenkin to Jenkins

File: workspace/locales/en.yml [hunk 4]
```diff
@@ -54,3 +54,13 @@ en:
           error: "Cannot find build %{buildtype} (%{buildnumber})!"
           list: "Here are the artifacts of %{build}:\n%{artifacts}"
           empty: "No artifact was found in the given build number (%{buildnumber})!"
+      jenkins:
+        help:
+          build_jenkins:
+            syntax: "Jenkin: `build_jenkin` `job_name` [`params`]"
+            desc: "Build now a job on Jenkin"
```
Comment (nguyentranlogigear) [inline workspace/locales/en.yml:61]: Please update Jenkin to Jenkins like Scopp's comment

File: workspace/lib/jenkinshelper/jenkins.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+          jobs << job["name"] if job["name"] =~ /#{filter}/i
+        else
+          jobs << job["name"] if job["name"] =~ /#{filter}/
+        end
+      end
+      jobs
+    end
+
+    def job_exists?(job_name)
+        list(job_name).include?(job_name)
```
Comment (nguyentranlogigear) [inline workspace/lib/jenkinshelper/jenkins.rb]: Could you help me align this line (4 space to 2 space)

File: workspace/lib/lita/handlers/jenkins.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        response.reply(t('build_jenkins.building', job_name: job_name))
+        begin
+          if job_exists?(job_name)
+            if (params.nil? or params.empty?)
+              client.api_post_request("/job/#{path_encode job_name}/build", {})
+            else
+              client.api_post_request("/job/#{path_encode job_name}/buildWithParameters", params)
+            end
+            queue_result = get_item_in_queue(job_name)
+            unless queue_result == false
```
Comment (nguyentranlogigear) [inline workspace/lib/lita/handlers/jenkins.rb:31]: Suggest: We can edit it 
<code>if queue_result</code>

Comment (nguyentunglam912) [reply workspace/lib/lita/handlers/jenkins.rb:31]: I want the expectation path executed first, then the exception

File: workspace/lib/jenkinshelper/jenkins.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
+require 'uri'
+module JenkinsHelper
+  module Jenkins
+    def form_encode(string)
+      URI.encode_www_form_component string.encode(Encoding::UTF_8)
+    end
```
Comment (vovuong) [inline workspace/lib/jenkinshelper/jenkins.rb]: I see that the `form_encode` is not used. Please remove if it is redundant.

File: workspace/lib/jenkinshelper/jenkins.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    def path_encode(path)
+      URI.escape(path.encode(Encoding::UTF_8))
+    end
+
+    def client
+      JenkinsApi::Client.new(
+      server_ip: config.ip,
+      username: config.username,
+      password: config.password)
+    end
```
Comment (vovuong) [inline workspace/lib/jenkinshelper/jenkins.rb:16]: Please align it as following format
JenkinsApi::Client.new(
_   server_ip: config.ip,
_   username: config.username,
_   password: config.password
)

File: workspace/lib/jenkinshelper/jenkins.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      response_json = client.api_get_request("")
+      jobs = []
+      response_json["jobs"].each do |job|
+        if ignorecase
+          jobs << job["name"] if job["name"] =~ /#{filter}/i
+        else
+          jobs << job["name"] if job["name"] =~ /#{filter}/
+        end
+      end
+      jobs
```
Comment (vovuong) [inline workspace/lib/jenkinshelper/jenkins.rb]: Should add `return` here to clarify.

File: workspace/lib/jenkinshelper/jenkins.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+    def job_exists?(job_name)
+      list(job_name).include?(job_name)
```
Comment (vovuong) [inline workspace/lib/jenkinshelper/jenkins.rb]: Should add `return` here to clarify.

**Comment by vovuong** [comment]
Please add buildbot image into description to show how it work?

**Pull request closed without merging**